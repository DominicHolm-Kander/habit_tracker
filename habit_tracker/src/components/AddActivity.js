import React from 'react'

export default function AddActivity() {
    return (
        <div>
            <form>
                    <label for="activityName">Name of activity
                    <input type="text" name="activityName"></input>
                    </label>
                    <label for="frequency">Frequency
                    <input type="text" name="frequency"></input>
                    </label>
                    <label for="priority">Priority (probs actually a radio button)
                    <input type="text" name="priority"></input>
                    </label>
                    <button type="button" id="addActivityButton" >Add an activity! (jsx needs updating)</button>
                </form>
        </div>
    )
}
