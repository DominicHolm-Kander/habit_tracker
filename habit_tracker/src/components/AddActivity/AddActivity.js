import React, { Component } from 'react'

export class AddActivity extends Component {
    state = {
        frequency: "Daily",
        isPriority: false
    }

    handleChangeOfActivityName = (event) => {
        this.setState({activityName: event.target.value})
    }
    handleChangeOfFrequency = (event) => {
        this.setState({frequency: event.target.value})
    }
    
    togglePriority = () => {
        if(this.state.isPriority){
            this.setState({isPriority : false})
        } else { 
            this.setState({isPriority: true})
        }
    }
    
    render() {
        return (
            <div>
            <form>
                <label for="activityName">Name of activity
                <input type="text" name="activityName" 
                onChange = {(event) => this.handleChangeOfActivityName(event)}>
                </input>
                </label>
                <label for="frequency">Frequency
                <select name="frequency" onChange = {(event) => this.handleChangeOfFrequency(event)}>
                    <option value = "Daily">Daily</option>
                    <option value = "Weekly">Weekly</option>
                    <option value = "Monthly">Monthly</option>
                </select>
                </label>
                <label for="priority">Priority (probs actually a radio button)
                <input type="checkbox" name="priority" onClick = {() => this.togglePriority()}/>This task is a priority
                </label>
                <button type="button" id="addActivityButton" onClick={() => this.props.addNewActivity(this.state)}>Add an activity! (jsx needs updating)</button>
            </form>
        </div>
        )
    }
}

export default AddActivity
