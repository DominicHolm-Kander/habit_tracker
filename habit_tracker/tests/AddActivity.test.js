import React from 'react';
import { shallow } from 'enzyme'
import AddActivity from '../src/components/AddActivity/AddActivity';
import waitUntil from "async-wait-until"; // For async testing of setState functions.
import regeneratorRuntime from 'regenerator-runtime';

describe("Shallow AddActivity", () => {
    let wrapper;
    let addActivityMock = jest.fn()

    beforeEach(() => wrapper = shallow(<AddActivity addNewActivity={addActivityMock} />));

    it("Should render correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });
    it("Should render 1 input element and 1 textarea", () => {
        expect(wrapper.find("input").length).toEqual(1);
        expect(wrapper.find("textarea").length).toEqual(1);
    })
    it("Should render a select element and a submit button", () => {
        expect(wrapper.find("select").length).toEqual(1);
        expect(wrapper.find("button").length).toEqual(1);
    })
    it("Submit button should call addNewActivity function", () => {
        expect(addActivityMock).not.toHaveBeenCalled();
        wrapper.find('button').simulate('click');
        expect(addActivityMock).toHaveBeenCalled();
    })
    it("HandleChangeOfActivityName function should update component state", async () => {
        wrapper.instance().handleChangeOfActivityName('testActivityName');
        await waitUntil(() => wrapper.instance().state.activityName);
        expect(wrapper.instance().state.activityName).toEqual('testActivityName');
    })
    it("HandleChangeOfFrequency function should update component state", async () => {
        wrapper.instance().handleChangeOfFrequency('testFrequencyValue');
        await waitUntil(() => wrapper.instance().state.frequency);
        expect(wrapper.instance().state.frequency).toEqual('testFrequencyValue');
    })
    it("togglePriority function should update component state, alternating between true and false",
    async () => {
        expect(wrapper.instance().state.isPriority).toBeFalsy();
        wrapper.instance().togglePriority();
        await waitUntil(() => wrapper.instance().state.isPriority);
        expect(wrapper.instance().state.isPriority).toEqual(true);
        wrapper.instance().togglePriority();
        await waitUntil(() => !wrapper.instance().state.isPriority);
        expect(wrapper.instance().state.isPriority).toBeFalsy();
    })
    it("Typing in the activityName input element should call handleChangeOfActivityName function", 
    () => {
        let spy = jest.spyOn(wrapper.instance(), "handleChangeOfActivityName");
        wrapper.instance().forceUpdate();  
        expect(spy).toHaveBeenCalledTimes(0);
        wrapper.find('[name="activityName"]').at(0)
        .simulate("change", { target: { value: "anything"}});
        expect(spy).toHaveBeenCalledTimes(1)
    })
    it("Selecting a frequency should call handleChangeOfFrequency function", 
    () => {
        let spy = jest.spyOn(wrapper.instance(), "handleChangeOfFrequency");
        wrapper.instance().forceUpdate();  
        expect(spy).toHaveBeenCalledTimes(0);
        wrapper.find('[name="frequency"]').at(0)
        .simulate("change", { target: { value: "Weekly"}});
        expect(spy).toHaveBeenCalledTimes(1)
    })
    it("Clicking the priority checkbox should call togglePriority function", 
    () => {
        let spy = jest.spyOn(wrapper.instance(), "togglePriority");
        wrapper.instance().forceUpdate();  
        expect(spy).toHaveBeenCalledTimes(0);
        wrapper.find('[name="priority"]').at(0)
        .simulate("click");
        expect(spy).toHaveBeenCalledTimes(1)
    })
})