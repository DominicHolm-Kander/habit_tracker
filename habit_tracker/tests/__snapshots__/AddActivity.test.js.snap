// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shallow AddActivity Should render correctly 1`] = `
<div
  className="container"
>
  <form>
    <div
      className="form-group"
    >
      <label
        for="exampleFormControlTextarea1"
      >
        What do you need to do?
      </label>
      <textarea
        className="form-control"
        id="exampleFormControlTextarea1"
        name="activityName"
        onChange={[Function]}
        rows="2"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        for="exampleFormControlSelect1"
      >
        Frequency
      </label>
      <select
        className="form-control"
        id="exampleFormControlSelect1"
        name="frequency"
        onChange={[Function]}
      >
        <option
          value="Daily"
        >
          Daily
        </option>
        <option
          value="Weekly"
        >
          Weekly
        </option>
        <option
          value="Monthly"
        >
          Monthly
        </option>
      </select>
    </div>
    <div
      className="form-group form-check"
    >
      <input
        className="form-check-input"
        id="exampleCheck1"
        name="priority"
        onClick={[Function]}
        type="checkbox"
      />
      <label
        className="form-check-label"
        for="exampleCheck1"
      >
        Is this habit important?
      </label>
    </div>
    <button
      className="btn btn-primary"
      id="addActivityButton"
      onClick={[Function]}
      type="button"
    >
      Add
    </button>
  </form>
</div>
`;
